### DISCUSSION TITLE
When running composer require I noticed that 9 packages were removed

### DISCUSSION THREAD

**Discussion by pildit** [composer/composer#10330] [Q&A / Support]
I just ran `composer require barry/debugbar` and it removed 9 packages !

Can anyone explain why this happened ?  

Here is the output :  

```
Using version ^3.6 for barryvdh/laravel-debugbar
./composer.json has been updated
Running composer update barryvdh/laravel-debugbar
Loading composer repositories with package information
Updating dependencies
Lock file operations: 3 installs, 0 updates, 9 removals
  - Removing asm89/stack-cors (v2.0.3)
  - Removing doctrine/annotations (1.13.2)
  - Removing friendsofphp/php-cs-fixer (v3.3.2)
  - Removing fruitcake/laravel-cors (v2.0.4)
  - Removing php-cs-fixer/diff (v2.0.2)
  - Removing psr/cache (1.0.1)
  - Removing symfony/options-resolver (v5.4.0)
  - Removing symfony/polyfill-php81 (v1.23.0)
  - Removing symfony/stopwatch (v5.4.0)
  - Locking barryvdh/laravel-debugbar (v3.6.4)
  - Locking maximebf/debugbar (v1.17.3)
  - Locking symfony/debug (v4.4.31)
Writing lock file
Installing dependencies from lock file (including require-dev)
Package operations: 3 installs, 0 updates, 9 removals
  - Removing symfony/stopwatch (v5.4.0)
  - Removing symfony/polyfill-php81 (v1.23.0)
  - Removing symfony/options-resolver (v5.4.0)
  - Removing psr/cache (1.0.1)
  - Removing php-cs-fixer/diff (v2.0.2)
  - Removing fruitcake/laravel-cors (v2.0.4)
  - Removing friendsofphp/php-cs-fixer (v3.3.2)
  - Removing doctrine/annotations (1.13.2)
  - Removing asm89/stack-cors (v2.0.3)
  - Installing symfony/debug (v4.4.31): Extracting archive
  - Installing maximebf/debugbar (v1.17.3): Extracting archive
  - Installing barryvdh/laravel-debugbar (v3.6.4): Extracting archive`
```

**Comment by stof** [comment]
One guess is that your lock file was outdated before, containing packages that were not actually required anymore (this was happening in some cases in Composer 1, or it could happen if you removed some stuff from your composer.json before without running the updates on the lock file).
Another guess would be that you have a plugin installed messing up with the dependency resolution

**Reply by pildit** [reply]
In composer.lock file the removed packages were present. They were missing from the composer.json file.

Is there a flag to run `composer require` without removing anything ?

**Reply by Seldaek** [reply]
There is no way to avoid autoremoval in composer 2. If some dangling package is present it'll be removed. If you still need it the fix is to require it explicitly with a proper constraint and not rely on chance to keep stuff around :)

**Reply by pildit** [reply]
But I needed those packages. The problem is : why did composer thought those were "dangling" packages ??

**Reply by stof** [reply]
because you don't have a requirement for them in your composer.json

**Reply by sovainfo** [reply]
That is what --dry-run is for. You never know what happens, make sure you can agree with what composer intends to do.

Obviously your primary requirements are leading, their resolvement result in the list of packages to be recorded in the lock file. Which is used for removing / downloading packages from vendor. No longer being a dependency is the reason for removal. They don't need to be a primary(direct) requirement. Their requirement can come from any package. Their removal comes from the assessment the package no longer needs to be there. On the original environment you can use `depends` to find out where it is a dependency, including replace or provide declarations.
